server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesResponse.java

```diff
@@ -114,10 +123,42 @@ private static void writeField(StreamOutput out,
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.field("fields", responseMap);
+        builder.field(FIELDS_FIELD.getPreferredName(), responseMap);
         return builder;
     }
 
+    public static FieldCapabilitiesResponse fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<FieldCapabilitiesResponse, Void> PARSER =
+        new ConstructingObjectParser<>("field_capabilities_response", true,
+            a -> new FieldCapabilitiesResponse(
+                ((List<Tuple<String, Map<String, FieldCapabilities>>>) a[0]).stream()
+                    .collect(Collectors.toMap(Tuple::v1, Tuple::v2))));
+
+    static {
+        PARSER.declareNamedObjects(ConstructingObjectParser.constructorArg(), (p, c, n) -> {
+            Map<String, FieldCapabilities> typeToCapabilities = parseTypeToCapabilities(p, n);
+            return new Tuple<>(n, typeToCapabilities);
+        }, FIELDS_FIELD);
+    }
+
+    private static Map<String, FieldCapabilities> parseTypeToCapabilities(XContentParser parser, String name) throws IOException {
+        Map<String, FieldCapabilities> typeToCapabilities = new HashMap<>();
+
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.FIELD_NAME, token, parser::getTokenLocation);
```

Comment:
I wonder if these two `ensureExpectedToken` don't make parsing too strict, if so the insertion of random fields should make tests fail if configured properly.

Reply:
In tests, I set the 'excludes filter' to prevent adding random fields under the `fields` key. I assumed it wasn't okay to have malformed entries in that map or the one below it, as each key represents a field or type name: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-field-caps.html#_response_format

I'm not generally clear though on when we aim to be lenient in terms of parsing -- any suggestions/ guidance is much appreciated!

Reply:
I had another look at how you test this, and I think that this is all fine. The reason why we are lenient with responses is that we want to be able to parse responses from future versions of Elasticsearch, which may or may not change the format of the response in backwards compatible ways. If we are strict even backwards compatible changes end up breaking our parsing code. In this case though, I think that if these two `ensureExpectedToken` were to throw exception, it would be due to a breaking change, which we can't do much about.